### PR TITLE
test(e2e): real EQ_train integration test + resume differential

### LINKED ISSUE #108: E2E tests: test_train.py — real EQ_train subprocess test with resume differential (refs #104)
## Scope

Fourth child of #104.  Runs `EQ_train` via subprocess on tiny demo config for a few optimizer steps, asserts checkpoints exist and that the resume path actually works.

## What lands

- `tests/test_train.py` with:
  - Test that consumes `pretrained_model_dir` fixture, asserts `model_run_dir/checkpoints/last.ckpt` (or whatever our convention is) exists.
  - Config-roundtrip: the saved `resolved_config.yaml` can be loaded back and reflects any CLI overrides.
  - **Resume differential** (MEICAR idiom): train for N steps, resume from the last checkpoint for another M steps, compare against training for (N+M) steps directly — end-state should be equivalent (within tolerance) when using the same seed.  This catches silent resume bugs like the one #86 fixed.
  - Optional: assert that `do_resume=True` with a drifted config raises the validation error introduced in #94.

## Depends on

- #105 (foundation).
- Existing resume-config-drift check (#94) if we want the drifted-config assertion.

## Acceptance

- Demo config trains in under 60s on CPU.
- Resume-equivalence assertion would fail if `do_resume` silently reinitialized the model.
- No reliance on real data or full pretraining runs.

### PR DESCRIPTION
**Pull request by mmcdermott**
## Summary

Fourth PR in the #104 E2E testing effort.  Stacks on #112 → #111 → #110 → `dev`.  Adds real assertions for `EQ_train` on top of the `eq_trained_model_dir` fixture.

Closes #108.  Refs #104.

## Scope note (in response to Copilot review)

The overall PR diff includes commits from merged-in ancestor branches — notably PR #121's Codecov action v4→v5 bump + `fail_ci_if_error: true → false` — that were merged into `dev` and then merged forward into this branch to keep it current.  **Those are ancestor history, not content this PR introduces.**  The per-file diff against the immediate parent branch shows the actual content added by this PR.  Called out explicitly here so the scope stays clear during review.

## What this PR adds

`tests/test_train.py` with 3 tests:

1. **`test_resolved_config_captures_overrides`** — asserts `resolved_config.yaml` records the CLI-override `output_dir` / `task_labels_dir` / `tensorized_cohort_dir` values, not the `???` sentinels.  Downstream tools (evaluate, reproduce-run scripts) depend on this being faithful.

2. **`test_checkpoint_metadata`** — Lightning checkpoint has the expected training-state keys (`state_dict`, `optimizer_states`, `global_step`, `epoch`, `hyper_parameters`), and `global_step` matches the demo config's `max_steps=2`.  State-dict sanity: contains ModernBERT backbone keys.

3. **Differential — `test_resume_advances_global_step`** (MEICAR idiom).  Stages the fixture's output into a fresh dir, runs `EQ_train` with `do_resume=True do_overwrite=False trainer.max_steps=4`, asserts the highest-`global_step` checkpoint advanced from 2 to 4.  **Strengthened per review**: also compares `state_dict` tensors element-wise and asserts at least one parameter changed — catches zero-lr / frozen-params / silent-no-op regressions that the global_step-only check would pass.

## Implementation note — Lightning checkpoint versioning

The resume test uses `_highest_step_checkpoint()` rather than loading `last.ckpt` directly.  When `ModelCheckpoint(save_last=True)` resumes into a dir that already has a `last.ckpt`, Lightning **does not overwrite** — it versions the new one as `last-v1.ckpt` and leaves the original in place.  Scanning all `*.ckpt` and picking the max-step one is robust to that behavior.

## Test plan

- [x] `uv run pytest tests/test_train.py -v` — 3 passed in ~29s
- [ ] CI green on push

## Merge order

Stacks behind #112 → #111 → #110.  Rebase chain forward as each parent merges.

### CODE REVIEW

**Comment by mmcdermott** [comment]
@copilot review

**Comment by mmcdermott** [comment]
@copilot review

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -34,11 +34,15 @@ jobs:
...
         with:
           token: ${{ secrets.CODECOV_TOKEN }}
           files: coverage.xml
-          fail_ci_if_error: true
+          # codecov/codecov-action@v4.0.1 pinned exact produced intermittent EPIPE errors on
+          # the gpg-verification subprocess when Codecov's CDN throttled the uploader binary.
+          # v5 uses the new Codecov CLI (no embedded gpg) and the floating major-tag gives us
+          # patch-level fixes without a second pin-bump PR every time.
+          fail_ci_if_error: false
           verbose: true
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply .github/workflows/tests.yaml:54]: The Codecov action v5 / `fail_ci_if_error` change was introduced by PR #121 (merged to `dev` separately), not by this PR.  It shows up in the overall diff here because this branch periodically merges `dev` to stay current with its parent chain (#112 → #111 → #110 → dev), and that merge brings all the intervening dev history with it.  Added a "Scope note" to the PR description making this explicit so reviewers don't expect to see Codecov-related code in the per-file diff against the immediate parent branch.  No code change on this PR for the CI/coverage policy — that already shipped on dev.

File: tests/test_train.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    )
+    assert resumed_ckpt["global_step"] == 4, (
+        f"resume with max_steps=4 advanced to global_step={resumed_ckpt['global_step']}, expected 4"
+    )
+
+    # Weights-changed assertion: catches the failure mode where Lightning's trainer.fit is
+    # called with a resumed ckpt but no gradient updates actually happen (zero-lr bug, frozen
+    # parameters, etc.).  In that case global_step advances (Lightning increments it per batch
+    # regardless of whether optimizer.step fires) but the state_dict is bitwise-identical.
+    # Pick a representative backbone parameter and compare element-wise.
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_train.py]: You're right — applied the reworded comment in ece8166.  The weights-changed assertion is now positioned as a *complementary* check to the global_step assertions above (both would stall under a zero-lr / frozen-params regression; comparing parameter tensors directly is the more reliable "training changed the model" check) rather than as a workaround for wrong global_step semantics.

File: tests/test_train.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            "do_overwrite=False",
+            "do_resume=True",
+            "trainer.max_steps=4",
+        ],
+        env=ENSURE_ENV_PLACEHOLDERS,
+        timeout=300.0,
+    )
+
+    _, resumed_ckpt = _highest_step_checkpoint(resume_dir)
+    assert resumed_ckpt["global_step"] > starting_step, (
```
Comment (mmcdermott) [inline tests/test_train.py]: you don't need this assert and the next one. They are redundant.

File: tests/test_train.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        f"do_resume=True did not advance global_step "
+        f"(starting={starting_step}, after resume={resumed_ckpt['global_step']}).  "
+        f"Likely silent reinit-on-resume regression."
+    )
+    assert resumed_ckpt["global_step"] == 4, (
+        f"resume with max_steps=4 advanced to global_step={resumed_ckpt['global_step']}, expected 4"
+    )
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_train.py]: I do like the no-op resume and assert identical weights -- that is a truly distinct test that the two redundant other test caes.

File: tests/test_train.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    assert resolved.exists(), f"resolved_config.yaml not written to {eq_trained_model_dir}"
+    cfg = OmegaConf.load(resolved)
+    # output_dir override → must reflect the actual tmp run dir, not the ??? sentinel.
+    assert cfg.output_dir == str(eq_trained_model_dir), (
+        f"resolved_config.output_dir={cfg.output_dir!r} does not match actual run dir "
+        f"{eq_trained_model_dir!r}"
+    )
+    # task_labels_dir / tensorized_cohort_dir also came through overrides.
+    assert cfg.datamodule.config.task_labels_dir, "task_labels_dir missing from resolved config"
+    assert cfg.datamodule.config.tensorized_cohort_dir, "tensorized_cohort_dir missing from resolved config"
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_train.py]: Applied in f3c349d.  The test now takes the `eq_preprocessed_dataset` and `eq_sampled_tasks_dir` fixtures and asserts `cfg.datamodule.config.task_labels_dir == str(eq_sampled_tasks_dir)` / `cfg.datamodule.config.tensorized_cohort_dir == str(eq_preprocessed_dataset)`.  A dropped override that left `'???'` in place (truthy) would now fail the equality check with a diagnostic showing the `???` vs the expected fixture path.

File: tests/test_train.py (hunk 3 above)
Comment (mmcdermott) [reply tests/test_train.py]: Agreed — both removed in f3c349d.  Once I thought through it, `global_step > starting_step` was strictly weaker than `global_step == 4`, and neither actually distinguishes "resume happened" from "silently-dropped resume that ran fresh and happened to reach max_steps" (a fresh `max_steps=4` run also ends at `global_step=4`).  The test is now a two-stage differential anchored on parameter-tensor identity (no-op resume: every tensor bit-identical; advancing resume: at least one differs), which actually catches the regression the old asserts were trying to guard against.

File: tests/test_train.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_train.py]: Applied in f3c349d — structured as a no-op resume (`max_steps == starting_step`) asserting `global_step` unchanged + every parameter tensor bit-identical vs. the starting snapshot, then the advancing resume asserts at least one tensor differs.  Implementing this surfaced an adjacent bug in `train.py` where `best_model_score` is None on a no-op resume (no validation runs when max_steps is already reached) but the post-training log formatted it unconditionally with `:.2f` — bundled the None-guard fix into this commit since the test wouldn't run without it.

File: tests/test_train.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    for key in ("state_dict", "optimizer_states", "global_step", "epoch", "hyper_parameters"):
+        assert key in ckpt, f"checkpoint missing expected key {key!r}"
+    # Demo config has max_steps=2; global_step advances once per optimizer step.
+    assert ckpt["global_step"] == 2, (
+        f"demo training ran for {ckpt['global_step']} steps, expected 2 per _demo_train.yaml"
+    )
+    # State dict is non-empty and contains ModernBERT backbone parameters.
+    assert len(ckpt["state_dict"]) > 0
+    assert any("HF_model" in k or "model." in k for k in ckpt["state_dict"]), (
+        f"state_dict doesn't appear to contain the model backbone: keys={list(ckpt['state_dict'])[:3]}"
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_train.py]: Applied in f3c349d.  Kept the loose predicate but reworded the comment + assertion message to match what it's actually checking ("does the state_dict contain model-parameter keys at all"), with a note that validating the specific backbone-module tree isn't this test's job — that would belong in a model-wiring test.
